File: cpp/use_meshfunction.cpp
#include "use_meshfunction.h"
#include <dolfin/mesh/Cell.h>
#include <dolfin/function/FunctionSpace.h>
#include <dolfin/fem/GenericDofMap.h>
#include <dolfin/common/Timer.h>

std::set<la_index> get_dofs0(const Function& F,\
                             const std::vector<double>& contour_tol,\
                             const unsigned int width)
{
  Timer master_timer("Getting dofs0");
  master_timer.start();

  assert(F.value_rank() == 0);                    // only scalar valued funcs
  assert(contour_tol.size() == 2);                // check bounds
  assert(contour_tol[0] < contour_tol[1]);

  boost::shared_ptr<const Mesh> mesh(F.function_space()->mesh()); 
  CellFunction<bool> cell_f(*mesh, false);
  unsigned int gdim = mesh->geometry().dim();
  unsigned int tdim = mesh->topology().dim();

  Timer search_timer("First search0");
  search_timer.start();
  // find cells with F(midpoint) within tolerance
  Array<double> values(0);
  
  for(CellIterator cell(*mesh); !cell.end(); ++cell)
  {
    Array<double> x(gdim, cell->midpoint().coordinates());
    F.eval(values, x);
    if((values[0] > contour_tol[0]) and (values[0] < contour_tol[1]))
    {
      cell_f[*cell] = true; 
    }
  }
  search_timer.stop();

  // loop over already marked cells and mark their neigboring cells
  Timer connectivity_timer("Connectivity0");
  connectivity_timer.start();
  mesh->init(tdim, tdim);
  unsigned int N = mesh->num_cells();
  connectivity_timer.stop();
 
  Timer band_timer("Band build0");
  band_timer.start();
  for(unsigned int depth = 0; depth < width; depth++)
  {
    std::vector<bool> new_cells(N, false);
    for(CellIterator cell(*mesh); !cell.end(); ++cell)
    {
      if(cell_f[*cell])
      {
        const unsigned int* neighbors = cell->entities(tdim);
        const unsigned int n_neighbors = cell->num_entities(tdim);
        for(unsigned int i = 0; i < n_neighbors; i++)
        { 
          new_cells[neighbors[i]] = true;
        }
      }
    }
    cell_f.set_values(new_cells);
  }
  band_timer.stop();

  Timer extract_timer("Extract dofs0");
  extract_timer.start();
  // extract dofs of marked cells
  std::set<la_index> dofs;

  boost::shared_ptr<const GenericDofMap> dofmap(F.function_space()->dofmap());
  for(CellIterator cell(*mesh); !cell.end(); ++cell)
  {
    if(cell_f[*cell])
    {
      std::vector<la_index> cell_dofs = dofmap->cell_dofs(cell->index());
      dofs.insert(cell_dofs.begin(), cell_dofs.end());
    }
  }
  extract_timer.stop();

  master_timer.stop();
  return dofs;
}

File: python/use_meshfunction.py
"""
Given a scalar function in space V(mesh) you want to identify dofs that lie
in cells that are part of the band of width 2*w around the contour_value.

1) Find the cells that have the contour_value
2) Look through mesh marking the neighbors of contour_cell. Repeat for depth.
3) Extract dofs all marked cells
"""

from dolfin import *
from numpy import zeros, uintp

def get_dofs0(f, contour_tol, width):
  master_timer = Timer("Getting dofs0")
  master_timer.start()

  mesh = f.function_space().mesh()
  mesh_f = CellFunction("size_t", mesh, 0)

  # get cell with countour
  timer = Timer("First search0")
  timer.start()
  for cell in cells(mesh):
    M = cell.midpoint()
    if contour_tol[0] < f(M) < contour_tol[1]:
      mesh_f[cell] = 1
  timer.stop()
  
  # create a band around cells with countour
  D = mesh.topology().dim()
  N = mesh.num_cells()

  timer = Timer("Connectivity0")
  timer.start()
  mesh.init(D, D)
  timer.stop()

  timer = Timer("Band build0")
  timer.start()
  for depth in range(width):
    new_cells = zeros(N, dtype=uintp) 
    for cell in cells(mesh):
      if mesh_f[cell]:
        neighbors = cell.entities(D)
        for neighbor in neighbors:
          new_cells[int(neighbor)] = 1
   
    mesh_f.set_values(new_cells)
  timer.stop()

  # extract dofs
  timer = Timer("Extract dofs0")
  timer.start()
  dofmap = f.function_space().dofmap()
  dofs = []
  for cell in cells(mesh):
    if mesh_f[cell]:
      cell_dofs = dofmap.cell_dofs(cell.index())
      dofs.extend(cell_dofs)
  timer.stop()

  timer = Timer("Make unique0")
  timer.start()
  dofs = set(dofs) # remove the duplicit dofs
  timer.stop()

  master_timer.stop()

  return dofs
